modules/caddytls/zerosslissuer.go

```diff
@@ -17,44 +17,49 @@ package caddytls
 import (
 	"context"
 	"crypto/x509"
-	"encoding/json"
 	"fmt"
-	"io"
-	"net/http"
-	"net/url"
-	"strings"
-	"sync"
+	"strconv"
+	"time"
 
 	"github.com/caddyserver/certmagic"
-	"github.com/mholt/acmez/acme"
 	"go.uber.org/zap"
 
 	"github.com/caddyserver/caddy/v2"
+	"github.com/caddyserver/caddy/v2/caddyconfig"
 	"github.com/caddyserver/caddy/v2/caddyconfig/caddyfile"
 )
 
 func init() {
 	caddy.RegisterModule(new(ZeroSSLIssuer))
 }
 
-// ZeroSSLIssuer makes an ACME issuer for getting certificates
-// from ZeroSSL by automatically generating EAB credentials.
-// Please be sure to set a valid email address in your config
-// so you can access/manage your domains in your ZeroSSL account.
-//
-// This issuer is only needed for automatic generation of EAB
-// credentials. If manually configuring/reusing EAB credentials,
-// the standard ACMEIssuer may be used if desired.
+// ZeroSSLIssuer uses the ZeroSSL API to get certificates.
+// Note that this is distinct from ZeroSSL's ACME endpoint.
+// To use ZeroSSL's ACME endpoint, use the ACMEIssuer
+// configured with ZeroSSL's ACME directory endpoint.
 type ZeroSSLIssuer struct {
-	*ACMEIssuer
-
 	// The API key (or "access key") for using the ZeroSSL API.
-	// This is optional, but can be used if you have an API key
-	// already and don't want to supply your email address.
+	// REQUIRED.
 	APIKey string `json:"api_key,omitempty"`
 
-	mu     sync.Mutex
-	logger *zap.Logger
+	// How many days the certificate should be valid for.
+	// Only certain values are accepted; see ZeroSSL docs.
+	ValidityDays int `json:"validity_days,omitempty"`
+
+	// The host to bind to when opening a listener for
+	// verifying domain names (or IPs).
+	ListenHost string `json:"listen_host,omitempty"`
+
+	// If HTTP is forwarded from port 80, specify the
+	// forwarded port here.
+	AlternateHTTPPort int `json:"alternate_http_port,omitempty"`
+
+	// Use CNAME validation instead of HTTP.
+	CNAMEValidation *DNSChallengeConfig `json:"cname_validation,omitempty"`
```

Comment:
Hmm? I'm confused, how is it CNAME validation? What does propagation checks have to do with CNAMEs? Those are unrelated concepts, no?

Reply:
That's what ZeroSSL API calls it :man_shrugging: A CNAME record needs to be set... I am not sure why it is that way.

Reply:
But why is `propagation` config within that? Doesn't make sense to me.

Reply:
The CNAME still needs to be visible/propagated in order to be validated. I think?

Reply:
Followed up on Slack about this one, I wasn't quite satisfied.

TL;DR apparently ZeroSSL uses CNAME as a method of domain verification, similar to how ACME DNS uses the TXT record. Super weird that they don't just use TXT instead. But whatever.